web/static/js/components/subjects/Subjects.jsx

```diff
@@ -0,0 +1,240 @@
+        <TableColumn>{subject.phoneNumber}</TableColumn>
+      </TableRow>
+    )
+  }
+}
+
+class Subjects extends Component {
+  props: {
+    campaignId: number,
+    subjects: {
+      count: number,
+      items: Subject[],
+      editingSubject: ?SubjectParams,
+      limit: number,
+      page: ?number,
+      targetPage: number,
+      fetching: boolean,
+    },
+    collectionActions: {
+      fetchSubjects: (campaignId: number, limit: number, targetPage: number) => void,
+      changeTargetPage: (targetPage: number) => void,
+    },
+    itemActions: {
+      createSubject: (campaignId: number, subject: SubjectParams) => void,
+      updateSubject: (campaignId: number, subject: SubjectParams) => void,
+      editingSubjectCancel: () => void,
+      subjectEditing: (fieldName: string, value: string) => void,
+      editSubject: (subject: SubjectParams) => void,
+    }
+  }
+
+  closeSubjectFormModal() {
+    this.props.itemActions.editingSubjectCancel()
+  }
+
+  showSubjectForm() {
+    this.props.itemActions.editSubject({phoneNumber: '', registrationIdentifier: ''})
+  }
+
+  onEditPhoneNumber = (value) => this.onEditField('phoneNumber', value)
+  onEditRegistrationIdentifier = (value) => this.onEditField('registrationIdentifier', value)
+
+  onEditField(fieldName, value) {
+    this.props.itemActions.subjectEditing(fieldName, value)
+  }
+
+  goToPage(targetPage: number) {
+    this.props.collectionActions.changeTargetPage(targetPage)
+  }
+
+  fetchTargetPage() {
+    const { limit, targetPage } = this.props.subjects
+    this.props.collectionActions.fetchSubjects(this.props.campaignId, limit, targetPage)
+  }
+
+  createSubject() {
+    if (this.props.subjects.editingSubject != null) {
+      this.props.itemActions.createSubject(this.props.campaignId, this.props.subjects.editingSubject)
+    } else {
+      throw new Error("You can't create without editing a Subject")
+    }
+  }
+
+  updateSubject() {
+    if (this.props.subjects.editingSubject != null) {
+      this.props.itemActions.updateSubject(this.props.campaignId, this.props.subjects.editingSubject)
+    } else {
+      throw new Error("You can't update without editing a Subject")
+    }
+  }
+
+  editSubject(subject: Subject) {
+    this.props.itemActions.editSubject(((subject: any): SubjectParams)) // Subject should work with SubjectParams due compatibility
+  }
+
+  pageTitle() {
+    return 'Subjects!'
+  }
+
+  componentDidUpdate() {
+    const {
+      page,
+      targetPage,
+      fetching
+    } = this.props.subjects
+    if (page != targetPage && !fetching) {
+      this.fetchTargetPage()
+    }
+  }
+
+  componentDidMount() {
+    this.fetchTargetPage()
+  }
+
+  render() {
+    const {
+      editingSubject,
+      page,
+      items,
+      limit,
+      fetching,
+      count
+    } = this.props.subjects
+
+    const showDialog = editingSubject != null
+    let subjectForm = null
+    if (editingSubject != null) {
+      subjectForm = <SubjectForm
```

Comment:
Suggestion: extract the right hand side to a private function to make this render more readable.